### PR TITLE
Fairlight CMI IIx audio improvements

### PR DESCRIPTION
**Pull request by MooglyGuy**
- Nearly complete rewrite of CMI-01A "Channel Card" emulation [Ryan Holtz]
- Fixed a number of bugs in the 6840 PTM emulation [Ryan Holtz]

### CODE REVIEW

File: src/mame/fairlight/cmi01a.cpp [hunk 1]
```diff
@@ -186,499 +143,600 @@ void cmi01a_device::device_reset()
...
+	}
 }
 
-void cmi01a_device::set_eclk(bool eclk)
+void cmi01a_device::set_zx_flipflop_clock(const bool zx_ff_clk)
 {
-	bool old_eclk = m_eclk;
-	m_eclk = eclk;
+	if (zx_ff_clk == m_zx_ff_clk || !m_run)
+		return;
```
Comment (cuavas) [inline src/mame/fairlight/cmi01a.cpp:408]: Should this still update `m_zx_ff_clk` even if `m_run` is false?

File: src/devices/machine/6840ptm.cpp [hunk 2]
```diff
@@ -675,14 +777,23 @@ void ptm6840_device::set_gate(int idx, int state)
 
 void ptm6840_device::set_clock(int idx, int state)
 {
+	const bool old_clk = m_clk[idx];
 	m_clk[idx] = state;
+	const bool rising_edge = !old_clk && state;
+	if (rising_edge)
+	{
+		return;
+	}
```
[the remark this answers is not in the document]
Comment (MooglyGuy) [reply src/devices/machine/6840ptm.cpp:792]: It should be checking for an unchanging value, yes. I've added an early-out.

**Comment by Tafoid** [comment]
I have a few machines that come up in testing which have changed behaviors likely due to device changes in this commit.

cpu09 - skeleton/tavernie.cpp
cpu09-diff

duet16 - matsushita/duet16.cpp
duet16-diff

zwackery - midway/zwackery.cpp
zwackery-diff

If more is found, I will post.

**Comment by MooglyGuy** [comment]
I'll get to it sometime this week.

**Pull request merged**